apps/desktop/src-tauri/src/inference_thread.rs

```diff
@@ -51,77 +57,185 @@ fn clean_prompt(s: &str) -> String {
         .replace("\n<human>: ", "\n===\nhuman: ")
 }
 
-pub fn spawn_inference_thread(req: InferenceThreadRequest) {
-    std::thread::spawn(move || {
-        let guard = req.model_guard.read();
+// Perhaps might be better to clone the model for each thread...
+pub fn start_inference(req: InferenceThreadRequest) -> Option<JoinHandle<()>> {
+    println!("Starting inference ...");
+
+    // Need to clone it to have its own arc
+    let guard_clone = Arc::clone(&req.model_guard);
+    let guard = guard_clone.read();
+    let model = match &*guard {
+        Some(m) => m,
+        None => {
+            println!("Model locked, cannot be loaded");
+            return None;
+        }
+    };
+
+    let mut session = model.start_session(Default::default());
+
+    let raw_prompt = clean_prompt(req.completion_request.prompt.as_str());
+    let prompt = &raw_prompt;
+    let mut output_request = OutputRequest::default();
+
+    let inference_params = InferenceParameters {
+        // n_batch: 4,
+        // n_threads: 2,
+        ..Default::default()
+    };
+
+    // Manual tokenization if needed
+    // let vocab = model.vocabulary();
+    // let prompt_tokens: Vec<i32> = vocab
+    //     .tokenize(prompt, true)
+    //     .into_par_iter() // Flatten to create a parallel iterator over the tuples
+    //     .flatten() // Flatten the nested vectors
+    //     .map(|(_, tok)| tok)
+    //     .collect();
+
+    // for batch in prompt_tokens.chunks(inference_params.n_batch) {
+    //     model.evaluate(&mut session, inference_params, batch, &mut output_request);
+    //     for &tk in batch {
+    //         // Update the tokens for this session
+    //         session.tokens.push(tk);
+    //     }
+    // }
+
+    match session.feed_prompt::<Infallible>(
+        model.as_ref(),
+        &inference_params,
+        prompt,
+        &mut output_request,
+        |_| Ok(InferenceFeedback::Continue),
+    ) {
+        Ok(_) => {}
+        Err(e) => {
+            println!("Error while feeding prompt: {:?}", e);
+            return None;
+        }
+    }
+
+    let handle = spawn_inference_thread(req, inference_params, session, output_request);
+    Some(handle)
+}
 
+fn spawn_inference_thread(
+    req: InferenceThreadRequest,
+    inference_params: InferenceParameters,
+    mut session: llm::InferenceSession,
+    mut output_request: OutputRequest,
+) -> JoinHandle<()> {
+    println!("Spawning inference thread...");
+    let handle = actix_web::rt::task::spawn_blocking(move || {
+        let mut rng = ChaCha8Rng::seed_from_u64(42);
```

Comment:
guessing you switched to ChaCha8Rng because it's much faster?

Reply:
It's slightly faster, but the actual speed gain is by separating feed_token and inference_next_token. The first one is super slow and require top-level allocation, whereas the 2nd can be put into a nested thread